transport/mixer/src/channel.rs

```diff
@@ -0,0 +1,469 @@
+    static ref METRIC_QUEUE_SIZE: SimpleGauge =
+        SimpleGauge::new("hopr_mixer_queue_size", "Current mixer queue size").unwrap();
+    static ref METRIC_MIXER_AVERAGE_DELAY: SimpleGauge = SimpleGauge::new(
+        "hopr_mixer_average_packet_delay",
+        "Average mixer packet delay averaged over a packet window"
+    )
+    .unwrap();
+}
+
+/// Mixing and delaying channel using random delay function.
+///
+/// Mixing is performed by assigning random delays to the ingress timestamp of data,
+/// then storing the values inside a binary heap with reversed ordering (max heap).
+/// This effectively creates a min heap behavior, which is required to ensure that
+/// data is released in order of their delay expiration.
+///
+/// When data arrives:
+/// 1. A random delay is assigned
+/// 2. Data is stored in the heap with its release timestamp
+/// 3. The heap maintains ordering so items with earliest release time are at the top
+///
+/// The channel uses a single timer thread that is instantiated on the first
+/// timer reset and shared across all operations. This channel is **unbounded** by nature
+/// using the `capacity` in the configuration to solely pre-allocate the buffer.
+struct Channel<T> {
+    /// Buffer holding the data with a timestamp ordering to ensure the min heap behavior.
+    buffer: BinaryHeap<Reverse<DelayedData<T>>>,
+    timer: futures_timer::Delay,
+    waker: Option<std::task::Waker>,
+    sender_count: usize,
+    receiver_active: bool,
+    cfg: MixerConfig,
+}
+
+/// Error returned by the [`Sender`].
+#[derive(Debug, thiserror::Error)]
+pub enum SenderError {
+    /// The channel is closed due to receiver being dropped.
+    #[error("Channel is closed")]
+    Closed,
+
+    /// The mutex lock over the channel failed.
+    #[error("Channel lock failed")]
+    Lock,
+}
+
+/// Sender object interacting with the mixing channel.
+pub struct Sender<T> {
+    channel: Arc<Mutex<Channel<T>>>,
+}
+
+impl<T> Clone for Sender<T> {
+    fn clone(&self) -> Self {
+        let mut channel = self.channel.lock().unwrap();
+        channel.sender_count += 1;
+
+        Sender {
+            channel: self.channel.clone(),
+        }
+    }
+}
+
+impl<T> Drop for Sender<T> {
+    fn drop(&mut self) {
+        let mut channel = self.channel.lock().unwrap();
+        channel.sender_count -= 1;
+        if channel.sender_count == 0 && !channel.receiver_active {
+            channel.waker = None;
+        }
+    }
+}
+
+impl<T> Sender<T> {
+    /// Send one item to the mixing channel.
+    pub fn send(&self, item: T) -> Result<(), SenderError> {
+        let mut sender = self.clone();
+        sender.start_send_unpin(item)
+    }
+}
+
+impl<T> futures::sink::Sink<T> for Sender<T> {
+    type Error = SenderError;
+
+    fn poll_ready(self: std::pin::Pin<&mut Self>, _cx: &mut std::task::Context<'_>) -> Poll<Result<(), Self::Error>> {
+        let is_active = self.channel.lock().map_err(|_| SenderError::Lock)?.receiver_active;
```

Comment:
Potential performance drawback here is due to acquiring lock in the `poll_ready` and the immediately re-aquiring it again in `start_send`.

In fact, Senders can be active independently on the underlying Channel. So each Sender can have its own `is_closed` property and should be checking it, instead of checking the value behind the Mutex. The Channel becomes inactive only once all Senders are dropped, and therefore a Sender cannot be active while the Channel becomes inactive.

Reply:
This is true, atomics could be used for tracking that as well.

Reply:
That is true. I will incorporate the change.